### ISSUE TITLE
`systemctl status` (with journald access) seems unnecessarily slow (disk bound)

### ISSUE THREAD

**Issue by sourcejedi** [systemd/systemd#7963]
### Submission type

 - Request for enhancement (RFE)

### systemd version the issue has been seen with

236 git ish

### Used distribution

Fedora 27

### In case of bug report: Expected behaviour you didn't see

> …

### In case of bug report: Unexpected behaviour you saw

Using `systemctl status`, which shows the last five log messages, takes over 10 times longer than `journalctl -b`, which already shows all the same lines.  This is particularly noticeable if you ask for several units; `systemctl status *.mount` takes 140x as long. 

I think it would be better to optimize `systemctl status`, to explicitly not request messages from previous boots.  IMO this would fit with the mental model of what `systemctl status` is supposed to do.

### In case of bug report: Steps to reproduce the problem

```
$ time journalctl -b >/dev/null

real	0m0.101s
user	0m0.049s
sys	0m0.018s

$ time systemctl status dbus >/dev/null

real	0m4.670s
user	0m0.021s
sys	0m0.071s

$ time systemctl status *.mount>/dev/null

real	0m14.654s
user	0m0.075s
sys	0m0.214s
```

**Comment by bl33pbl0p** [comment]
Hmm, that's weird.
```
delta ~ $ time journalctl -b >/dev/null

real	0m0.329s
user	0m0.056s
sys	0m0.000s
delta ~ $ time systemctl status dbus > /dev/null

real	0m0.008s
user	0m0.003s
sys	0m0.002s
delta ~ $ time systemctl status *.mount>/dev/null

real	0m0.035s
user	0m0.005s
sys	0m0.013s
```

**Comment by sourcejedi** [comment]
Weird indeed!  I have noticed this for a while, that `systemctl status` takes a few seconds to run.

Obviously my system has a persistent journal (i.e. saved across boots, and not just kept in the tmpfs /run/systemd/journal), and it has been around for some while.

`du -ch /var/log/journal` shows 1.2G.  systemd-coredump isn't configured to dump into the journal, so surely it's all log messages...

ah, hum.  It looks like it might be an unfortunate result of the crash logic, however that works.  (Also, this is on ext4, but I'm wondering if journald is preallocating excessively to avoid more btrfs woes).

$ du -ah /var/log/journal/ec36f0d290614a989246ce31e09df42e | sort -h | head
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@00055529c4b41469-5fa2261e52fef98b.journal~
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@00055529d2c9e916-6bf27f09ca0cc120.journal~
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@0005554f9468681b-0997a8af3c04e933.journal~
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@0005554fb7e6d845-8d89ed69c313e8c2.journal~
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@00055588d25afba3-f5f61491f3e02caa.journal~
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@000555b6e406fa5f-201c4199819347ce.journal~
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@000555dd78843b30-13361ece58800741.journal~
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@0005564159c8456d-6ca5c86b254b0390.journal~
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@000556e20ed8efdb-4c97262997a6c143.journal~
8.1M	/var/log/journal/ec36f0d290614a989246ce31e09df42e/system@000556e2fa5b5c76-50fba37e5d66c3f3.journal~
$ du -ah /var/log/journal/ec36f0d290614a989246ce31e09df42e|wc -l
101

**Comment by bl33pbl0p** [comment]
XFS here, disk usage currently is 304.0M, coredump logs backtraces to the journal here, so not entirely log messages, and the journal is indeed persistent.

```
17M	/var/log/journal/cabbd0ab8b7742b5a135974616f592a1/user-1000@000561c8ff51f65e-9c409d5069741dce.journal~
17M	/var/log/journal/cabbd0ab8b7742b5a135974616f592a1/user-1000@36d8cc8a7b1f4d1fac80068be50f07a0-00000000000004c1-000561c8ff4d0096.journal
25M	/var/log/journal/cabbd0ab8b7742b5a135974616f592a1/system@000561c8fcbe39ec-a310bea3944a02d2.journal~
25M	/var/log/journal/cabbd0ab8b7742b5a135974616f592a1/user-1000.journal
32M	/var/log/journal/cabbd0ab8b7742b5a135974616f592a1/system@0005618a431687ed-c0e6fe776d47192b.journal~
81M	/var/log/journal/cabbd0ab8b7742b5a135974616f592a1/system.journal
113M	/var/log/journal/cabbd0ab8b7742b5a135974616f592a1/system@72510af5834c48c998a9088388519fb3-0000000000000001-000561c8fba14bd3.journal
305M	/var/log/journal/cabbd0ab8b7742b5a135974616f592a1/
```

**Comment by gdamjan** [comment]
it would make sense for `systemctl status <service>` to only show the logs from the current invocation (for running services) or from the last invocation for stopped services.

**Comment by sourcejedi** [comment]
[If I use `/usr/bin/time` instead, I can see e.g. 73632 `inputs` (of 4096 byte pages, so ~290MB) causing a delay of 2.3s](https://unix.stackexchange.com/a/420289/29483).  This sounds like expected for sequential IO on my hardware.  So I think we can focus on that `inputs` metric.

... and that was for `journalctl -b -u <service>`.  It seems my problem is wider than just `systemctl status`.

I think we actually already did what I suggested about `systemctl status` limiting it's output to the current boot.  Maybe I wasn't testing consistently, or maybe `-b -u <service>` is actually working slower than plain `-b`.  At least on my system.

https://github.com/systemd/systemd/blob/7755083256a89dfd8858cd62f1ee85c4b1327092/src/shared/logs-show.c#L1453
